core/services/assets/Registry.php

```diff
@@ -256,6 +293,73 @@ public function getData($key)
     }
 
 
+    /**
+     * Get the actual asset path for asset manifests.
+     * If there is no asset path found for the given $chunk_name, then the $chunk_name is returned.
+     * @param string $chunk_name
+     * @param string $asset_type
+     * @return string
+     * @throws InvalidFilePathException
+     * @since $VID:$
+     */
+    public function getAssetUrl($chunk_name, $asset_type)
+    {
+        if (empty($this->cached_manifest)) {
+            $this->registerManifests();
+        }
+        return isset($this->cached_manifest[$chunk_name][$asset_type])
+            ? $this->domain->distributionAssetsUrl() . $this->cached_manifest[$chunk_name][$asset_type]
+            : $chunk_name;
+    }
+
+
+    /**
+     * Used to register a js/css manifest file with the registered_manifest_files property.
+     *
+     * @param string $manifest_file  The absolute path to the manifest file.
+     * @throws InvalidFilePathException
+     * @since $VID:$
+     */
+    public function registerManifestFile($manifest_file)
+    {
+        if (! file_exists($manifest_file)) {
```

Comment:
this file exists check is repeated in registerManifests because if we keep the filter we still have to do a file_exists check.  So options are:

1. Keep things as is here.
2. Ditch the file_exists check in this method and just leave the one in `registerManifests`
3. Ditch the filter and file_exists check in `registerManifests`.  

Which would you prefer?  I'm on the fence.  I'm leaning a bit towards explicitly requiring manifest files to be registered using `registerManifestFile` myself.

Reply:
ya I would have `registerManifests()` simply call `registerManifestFile()` when looping over the returned array, then ALL of the validation logic and/or any other hooks/triggers can all reside in one place.

so 3 basically.

Reply:
Hmm... if you're suggesting ditch the filter, then I don't understand why `registerManifestFile` would get called from `registerManifests`?  If the filter is ditched then `registerManifests` will simply loop over the `registered_manifest_files` property which would have been already populated by calls to `registerManifestFiles` earlier in the request right?

Reply:
my apologies... i read the code wrong... I was getting confused between the `$registered_manifest_files` and `$cached_manifest` properties.

I think this could be simplified.

We don't need two arrays for tracking these files, one is sufficient imho.

Here's what I recommend:

 * get rid of the `$cached_manifest` property array as there is no need to shuffle data back n forth like we are doing now (wait for the rest of the explanation before coming to any conclusions)
 * have `registerManifestFile()` as a public method for adding files and have it perform ALL of the necessary validation and decoding of the file content. All valid decoded file content gets added to  `$registered_manifest_files` array (which could even just be renamed to `$manifest_files` for extra simplicity)
 * `registerManifests()` doesn't need to loop over any existing array of data, it can simply supply an empty array as its filtered property for add-ons to add their manifest files to. If the current approach was to allow the filter to deregister files, then that's fine, although this can easily be achieved by also adding the `$manifest_files` array property as a parameter on the filter. This would allow callbacks to edit the existing data if need be and probably provide better access to that data rather than having no option but to completely overwrite an existing manifest file (although I'll admit that I don't know what that content will look like so maybe this is a moo point).
 * so any code with access to the assets Registry class (like the Register Addon API one day) can call `registerManifestFile()` and directly register a file, or code can hook into the filter and add their file that way, which will also end up calling  `registerManifestFile()`.
 * then when `getAssetUrl()` is called, it will still call `registerManifests()`, which adds any *extra* manifest files to the `$manifest_files` array which `getAssetUrl()` uses to return its result.

does that make sense?

does the same thing, but is a little more simple and straightforward.